pkg/steps/source.go

```diff
@@ -539,20 +542,51 @@ func hintsAtInfraReason(logSnippet string) bool {
 		strings.Contains(logSnippet, "connection reset by peer")
 }
 
-func waitForBuildOrTimeout(ctx context.Context, buildClient BuildClient, namespace, name string) error {
-	return waitForBuild(ctx, buildClient, namespace, name)
+func waitForBuildOrTimeout(
+	ctx context.Context,
+	buildClient BuildClient,
+	podClient kubernetes.PodClient,
+	namespace, name string,
+) error {
+	return waitForBuild(ctx, buildClient, podClient, namespace, name)
 }
 
-func waitForBuild(ctx context.Context, buildClient BuildClient, namespace, name string) error {
+func waitForBuild(
+	ctx context.Context,
+	buildClient BuildClient,
+	podClient kubernetes.PodClient,
+	namespace, name string,
+) error {
 	logrus.WithFields(logrus.Fields{
 		"namespace": namespace,
 		"name":      name,
 	}).Trace("Waiting for build to be complete.")
-
-	evaluatorFunc := func(obj runtime.Object) (bool, error) {
-		switch build := obj.(type) {
-		case *buildapi.Build:
+	var ret atomic.Pointer[buildapi.Build]
+	var eg *errgroup.Group
+	eg, ctx = errgroup.WithContext(ctx)
+	pendingCtx, cancel := context.WithCancel(ctx)
+	pendingCheck := func() error {
+		timeout := podClient.GetPendingTimeout()
+		select {
+		case <-pendingCtx.Done():
+		case <-time.After(time.Until(ret.Load().CreationTimestamp.Add(timeout))):
+			if err := checkPending(ctx, podClient, ret.Load(), timeout, time.Now()); err != nil {
+				logrus.Infof(err.Error())
+				return err
+			}
+		}
+		return nil
+	}
+	eg.Go(func() error {
+		defer cancel()
+		return kubernetes.WaitForConditionOnObject(ctx, buildClient, ctrlruntimeclient.ObjectKey{Namespace: namespace, Name: name}, &buildapi.BuildList{}, &buildapi.Build{}, func(obj runtime.Object) (bool, error) {
+			build := obj.(*buildapi.Build)
+			first := ret.Swap(build) == nil
 			switch build.Status.Phase {
+			case buildapi.BuildPhaseNew, buildapi.BuildPhasePending:
```

Comment:
Can you expand?
- I don't understand the limit of 1, since the pending check must execute in parallel.
- We may miss the "new" phase if there is a delay between creating the build and establishing the watch, or if we are a separate execution entirely.
- This check must be started only once, while we may see several updates in the "new" phase.
- (If you purposefully did not include the shared pointer) there is no need for a separate `get` (we used to this for the pod watch) since we've already received the information from the watch and it is guaranteed to be the latest available.

Reply:
> ~~`eg.SetLimit(1)`~~

My bad, I should have written `eg.SetLimit(2)` instead, therefore the whole code would look like
```golang
eg.SetLimit(1)
// ...
switch build.Status.Phase {
case buildapi.BuildPhaseNew:
  _ = eg.TryGo(pendingCheck)
// ...
}
```
but it's maybe more hidden and ambiguous, so no way.
The rest of the explanation makes sense to me, thanks.

Reply:
I see, that makes sense.  That would be one way of having an extra flag indicating whether it has already started, as I said above.